atlasdb-config/src/main/java/com/palantir/atlasdb/http/VersionSelectingClients.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.http;
+
+import java.util.Map;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.DoubleSupplier;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.collect.ImmutableMap;
+import com.palantir.atlasdb.util.AtlasDbMetrics;
+import com.palantir.common.proxy.PredicateSwitchedProxy;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+
+/**
+ * Factory for randomly selecting from a new and a legacy version of a client, based on a live-reloading probability.
+ *
+ * Please note that your clients will run independently; if there are stateful invariants that need to be enforced
+ * across individual clients, you may need to share state appropriately.
+ */
+final class VersionSelectingClients {
+    private static final String CLIENT_VERSION = "clientVersion";
+    private static final Map<String, String> NEW_CLIENT_TAG = ImmutableMap.of(CLIENT_VERSION, "new");
```

Comment:
offline discussion about using versions here, probably also want to distinguish them from server versions maybe? unclear

Reply:
I'm not familiar, but doesn't this naming scheme introduce problems if there are more than two versions?

Reply:
Correct, this naming scheme would not work if you have >2 versions. I think the idea was that _clients_ would rarely have to work with >2 versions at a time, though our discussion was that it might still be useful to support more precise names, since if you had a service using `V1` or `V2`, and then upgraded to one with `V2` or `V3` (having the old one is useful in case you need to quickly stop the new one from running), then `V2` is new and then suddenly becomes old, and `AtlasVersion` isn't a tag that our metrics normally have so it's not very easy to unpick that immediately.